src/Storages/MergeTree/GinIndexStore.cpp

```diff
@@ -0,0 +1,447 @@
+        std::unique_ptr<char[]> buf(new char[size]);
+
+        buffer.readStrict(reinterpret_cast<char*>(buf.get()), size);
+
+        GinIndexPostingsListPtr postings_list = std::make_shared<GinIndexPostingsList>(GinIndexPostingsList::read(buf.get()));
+
+        return postings_list;
+    }
+}
+
+bool GinIndexStore::exists() const
+{
+    String id_file_name = getName() + ".gin_sid";
+    return storage->exists(id_file_name);
+}
+
+UInt32 GinIndexStore::getNextIDRange(const String& file_name, size_t n)
+{
+    std::lock_guard<std::mutex> guard{gin_index_store_mutex};
+
+    if (!storage->exists(file_name))
+    {
+        std::unique_ptr<DB::WriteBufferFromFileBase> ostr = this->data_part_storage_builder->writeFile(file_name, DBMS_DEFAULT_BUFFER_SIZE, {});
+
+        const auto& int_type = DB::DataTypePtr(std::make_shared<DB::DataTypeUInt32>());
+        auto size_serialization = int_type->getDefaultSerialization();
+        size_serialization->serializeBinary(1, *ostr);
+        ostr->sync();
+    }
+
+    /// read id in file
+    UInt32 result = 0;
+    {
+        std::unique_ptr<DB::ReadBufferFromFileBase> istr = this->storage->readFile(file_name, {}, std::nullopt, std::nullopt);
+
+        Field field_rows;
+        const auto& size_type = DB::DataTypePtr(std::make_shared<DB::DataTypeUInt32>());
+        auto size_serialization = size_type->getDefaultSerialization();
+
+        size_type->getDefaultSerialization()->deserializeBinary(field_rows, *istr);
+        result = field_rows.get<UInt32>();
+    }
+    //save result+n
+    {
+        std::unique_ptr<DB::WriteBufferFromFileBase> ostr = this->data_part_storage_builder->writeFile(file_name, DBMS_DEFAULT_BUFFER_SIZE, {});
+
+        const auto& int_type = DB::DataTypePtr(std::make_shared<DB::DataTypeUInt32>());
+        auto size_serialization = int_type->getDefaultSerialization();
+        size_serialization->serializeBinary(result + n, *ostr);
+        ostr->sync();
+    }
+    return result;
+}
+
+UInt32 GinIndexStore::getNextRowIDRange(size_t n)
+{
+    UInt32 result =current_segment.next_row_id;
+    current_segment.next_row_id += n;
+    return result;
+}
+
+UInt32 GinIndexStore::getNextSegmentID()
+{
+    String sid_file_name = getName() + ".gin_sid";
+    return getNextIDRange(sid_file_name, 1);
+}
+
+UInt32 GinIndexStore::getSegmentNum()
+{
+    String sid_file_name = getName() + ".gin_sid";
+    if (!storage->exists(sid_file_name))
+        return 0;
+    Int32 result = 0;
+    {
+        std::unique_ptr<DB::ReadBufferFromFileBase> istr = this->storage->readFile(sid_file_name, {}, std::nullopt, std::nullopt);
+
+        Field field_rows;
+        const auto& size_type = DB::DataTypePtr(std::make_shared<DB::DataTypeUInt32>());
+        auto size_serialization = size_type->getDefaultSerialization();
+
+        size_type->getDefaultSerialization()->deserializeBinary(field_rows, *istr);
+        result = field_rows.get<UInt32>();
+    }
+    return result - 1;
```

Comment:
Hmm, why `- 1`?

Reply:
Because the written result is the next available segment ID, so the segment number is the value minus 1.